libzmix/src/utils/bls12_381/mod.rs

```diff
@@ -0,0 +1,211 @@
+                    }
+                }
+
+                deserializer.deserialize_str($visitor)
+            }
+        }
+    };
+}
+
+macro_rules! format_impl {
+    ($name:ident) => {
+        impl std::fmt::Display for $name {
+            fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
+                write!(f, "$name( {} )", self.to_hex())
+            }
+        }
+
+        impl std::fmt::Debug for $name {
+            fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
+                write!(f, "$name( {} )", self.to_hex())
+            }
+        }
+    };
+}
+
+/// Generate a random number twice the number of bytes as the curve reduced by mod CURVE_ORDER
+/// Here 96 bytes are randomly generated mod CURVE_ORDER. The result is 48 bytes.
+fn random_mod_order<R: RngCore>(r: Option<&mut R>) -> BIG {
+    let mut seed1 = vec![0u8; rom::MODBYTES];
+    let mut seed2 = vec![0u8; rom::MODBYTES];
+
+    match r {
+        Some(rr) => {
+            rr.fill_bytes(&mut seed1.as_mut_slice());
+            rr.fill_bytes(&mut seed2.as_mut_slice());
+        }
+        None => {
+            thread_rng().fill_bytes(&mut seed1.as_mut_slice());
+            thread_rng().fill_bytes(&mut seed2.as_mut_slice());
+        }
+    }
+
+    compute_big(seed1.as_slice(), seed2.as_slice())
+}
+
+/// Hash the data to a field order element
+/// Based on standardization efforts for BLS12-381, it is recommended to use HKDF
+/// instead of SHA2 due to its properties of better hiding low entropy inputs.
+///
+/// Here, 96 bytes is used. This allows a bigger range than SHA2-256 and eliminates some biases
+/// for outputs that are close to the CURVE_ORDER like SHA2-384.
+///
+/// Salt should be of sufficient length to add the security of the keying material (See https://tools.ietf.org/html/rfc5869)
+/// The output of 96 bytes should require the salt to not be less than 32 bytes
+///
+/// The result is 48 bytes.
+fn hash_mod_order(data: &[u8], salt: &[u8], domain_sep_context: &[u8]) -> BIG {
+    //Domain salt and key
+    let hk = Hkdf::<Sha256>::extract(Some(salt), data);
+    let mut output = vec![0u8; rom::MODBYTES * 2];
+    hk.expand(domain_sep_context, output.as_mut_slice())
+        .unwrap();
+
+    compute_big(
+        &output.as_slice()[0..rom::MODBYTES],
+        &output.as_slice()[rom::MODBYTES..],
+    )
+}
+
+fn compute_big(seed1: &[u8], seed2: &[u8]) -> BIG {
+    let num1 = BIG::frombytes(seed1);
+    let num2 = BIG::frombytes(seed2);
+    let num1 = DBIG::new_scopy(&num1);
+    let mut res = DBIG::new();
+    res.ucopy(&num2);
+    res.add(&num1);
+    res.dmod(&GROUP_ORDER)
+}
+
+pub mod fr;
+
+#[cfg(test)]
+mod tests {
+    use super::{hash_mod_order, random_mod_order};
+    use amcl::bls381::big::BIG;
+    use amcl::bls381::rom;
+    use rand::prelude::*;
+    use rand_chacha::ChaChaRng;
+    use rand_core::SeedableRng;
+
+    #[test]
```

Comment:
As above, I'd like to see these tests be broken up into a number of different ones, and negative tests added. 
Should there be a test to ensure that the output size is 48 bytes, or that the salt is not less than 32 bytes?

Reply:
I agree with the test for output size.